### ISSUE TITLE
SSL login via Thunderbird -- certificate doesn't match, despite following exact code from Wiki

### ISSUE THREAD

**Issue by xmatic** [tomav/docker-mailserver#1723]
## Subject
* [ ] I have questions about TLS/SSL/STARTTLS/OpenSSL

## Description

My mailserver works. Am able to send and receive email. 
However, the `SSL` login doesn't work.
I used the exact code from here: https://github.com/tomav/docker-mailserver/wiki/Configure-SSL#lets-encrypt-recommended

However, when I try to log-into the email via Thunderbird, it tries to go to `imap.example.com:143` and says the certificate is mismatched. Here are 2 screencaps from Thunderbird's process:

(In the code below, `example.com` is replaced with my actual `domain.tld` in the actual code, and the email address is `mail@example.com`).

![2020-12-19_031101](https://user-images.githubusercontent.com/5103936/102687406-16f0eb00-41ac-11eb-9366-1bfefea01d6b.jpg)

![2020-12-19_031614](https://user-images.githubusercontent.com/5103936/102687419-2f610580-41ac-11eb-9259-4c31618bc7f9.jpg)

In the first screenshot, you can see that Thunderbird did some auto-configuration.

In the second screenshot, it's saying that the certificate for `imap.example.com:143` isn't as expected.

Thing is, nowhere in the process do I indicate `imap.example.com` or `smtp.example.com` to the `Lets Encrypt bot`. I indicate `mail.example.com` in the `.env` file. But not `imap.example.com` or `smtp.example.com`. 

**Shouldn't I be indicating `imap.example.com` and `smtp.example.com` somewhere?**

-------------------------------------------------

All my code is reproduced below:

$ `sudo docker network create nginx-proxy`

Then I navigate to the `sidecar` and `mailserver` directories and do: `sudo docker-compose up -d`

### sidecar/docker-compose.yml
```
version: "2"

services:
  nginx: 
    image: nginx
    container_name: nginx
    ports:
      - 80:80
      - 443:443
    volumes:
      - /mnt/data/nginx/htpasswd:/etc/nginx/htpasswd
      - /mnt/data/nginx/conf.d:/etc/nginx/conf.d
      - /mnt/data/nginx/vhost.d:/etc/nginx/vhost.d
      - /mnt/data/nginx/html:/usr/share/nginx/html
      - /mnt/data/nginx/certs:/etc/nginx/certs:ro
    networks:
      - proxy-tier
    restart: always

  nginx-gen:
    image: jwilder/docker-gen
    container_name: nginx-gen
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - /mnt/data/nginx/templates/nginx.tmpl:/etc/docker-gen/templates/nginx.tmpl:ro
    volumes_from:
      - nginx
    entrypoint: /usr/local/bin/docker-gen -notify-sighup nginx -watch -wait 5s:30s /etc/docker-gen/templates/nginx.tmpl /etc/nginx/conf.d/default.conf
    restart: always

  letsencrypt-nginx-proxy-companion:
    image: jrcs/letsencrypt-nginx-proxy-companion
    container_name: letsencrypt-companion
    volumes_from:
      - nginx
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /mnt/data/nginx/certs:/etc/nginx/certs:rw
    environment:
      - NGINX_DOCKER_GEN_CONTAINER=nginx-gen
      - DEBUG=false
    restart: always

networks:
  proxy-tier:
    external:
      name: nginx-proxy
```

### mailserver/docker-compose.yml
```
version: '2'
services:
  mail:
    image: tvial/docker-mailserver:latest
    hostname: ${HOSTNAME}
    domainname: ${DOMAINNAME}
    container_name: ${CONTAINER_NAME}
    ports:
    - "25:25"
    - "143:143"
    - "465:465"
    - "587:587"
    - "993:993"
    volumes:
    - ./mail:/var/mail
    - ./mail-state:/var/mail-state
    - ./config/:/tmp/docker-mailserver/
    - /mnt/data/nginx/certs/:/etc/letsencrypt/live/:ro
    cap_add:
    - NET_ADMIN
    - SYS_PTRACE
    restart: always

  cert-companion:
    image: nginx
    environment:
      - "VIRTUAL_HOST=mail.example.com"
      - "VIRTUAL_NETWORK=nginx-proxy"
      - "LETSENCRYPT_HOST=mail.example.com"
      - "LETSENCRYPT_EMAIL=example+letsencrypt@gmail.com"
    networks:
      - proxy-tier
    restart: always
    
networks:
  proxy-tier:
    external:
      name: nginx-proxy
```

### mailserver/.env

```
# –––––––––––––––––––––––––––––––––––––––––––––––
# ––– Docker Compose Settings Variables –––––––––
# –––––––––––––––––––––––––––––––––––––––––––––––

HOSTNAME=mail
DOMAINNAME=example.com
CONTAINER_NAME=mail      #can't use container name if there's name clash with another container

# –––––––––––––––––––––––––––––––––––––––––––––––
# ––– SELinux Compose File Settings Variables –––
# –––––––––––––––––––––––––––––––––––––––––––––––

SELINUX_LABEL=-Z

```

**Comment by aendeavor** [comment]
From the screenshot I see that the protocol field is empty? Moreover, the username you're using is probably wrong: For me, it's: `example@my-domain.com` and not `example`. In your case thats `example@${DOMAINNAME}.com`. But I do not think this is the main issue here. More plausible is that your Let'sEncrypt certificates are not valid for the subdomains Thunderbird chose, which are `imap.${DOMAINNAME}.com` and `smtp.${DOMAINNAME}.com` for IMAP and SMTP respectively. Not sure what NGINX does here reverse-proxy wise, but your IMAP and SMTP domain are probably the same.

**Comment by xmatic** [comment]
> From the screenshot I see that the protocol field is empty? Moreover, the username you're using is probably wrong: For me, it's: `example@my-domain.com` and not `example`. In your case thats `example@${DOMAINNAME}.com`. But I do not think this is the main issue here. More plausible is that your Let'sEncrypt certificates are not valid for the subdomains Thunderbird chose, which are `imap.${DOMAINNAME}.com` and `smtp.${DOMAINNAME}.com` for IMAP and SMTP respectively. Not sure what NGINX does here reverse-proxy wise, but your IMAP and SMTP domain are probably the same.

@aendeavor Thank you for replying. I added the following line near the top of my question:

> (In the code below, `example.com` is replaced with my actual `domain.tld` in the actual code, and the email address is `mail@example.com`).

**Comment by aendeavor** [comment]
Okay, let's get through this one by one:

---

1. > Final question: What client are you using please?

Definitely Thunderbird. In my opinion the only valid option. I'm using it with docker-mailserver and it works like a charm! GPG and S/MIME work great too. Stay with Thunderbird. It is not the problem here.

2. I noticed some oddities: Let's start with DNS. Who is authoritative DNS here? The picture says: "This domain is using third-party nameservers. DNS added here won't have an effect... ". Make sure you are using the DNS provider that has the [authoritative DNS](https://www.cloudns.net/blog/authoritative-dns-server/) for your domain. Moreover two records seem odd: The `* IN A 104...` record. I have never seen this and can only image what you are trying to do - but DNS is not meant to be used this way. I'd remove this record as fast as possible. The other records seem fine, except the A record for `mail.${DOMAINNAME}`. Does this really need to be an `A` record, or does it point to the same domain as the second A record, ie. `IN A @ 104...`? If so, use a `CNAME`.

3. > I see what you mean by protocol field. Have looked into it. Strangely, after googling there seems to be nothing online on how to fill this field.

Then it's not important:)

4. The manually configured settings seem to be correct. This is however expected, since I already thought it was a certificate problem.

5. > I agree that Thunderbird is looking for a host at: imap.example.com. Is it supposed to find imap at mail.example.com:143 instead? 

No, definitely not. The problem _was_ that `imap.${DOMAINNAME}` does not exist according to your DNS and it is not important. Thunderbird assumes there are different subdomains for IMAP and SMTP, but this is not the case here. I don't use it either. If you mailserver is reached on `mail.$DOMAINNAME`, it should be fine.

---

Which leads us to the (from my PoV) only valid problem here: certificates. From experience I'd say either you mounted them incorrectly, or your certificates are not valid for the `mail.${DOMAINNAME}` subdomain. I'm not an NGINX expert, but I'd `exec` into the mailserver and first of all look whether they are really there (with `docker exec mail bash` if you're using Compose).

> For the first command I got:
> ```
> unable to load certificate
> 139899794376000:error:0909006C:PEM routines:get_name:no start line:../crypto/pem/pem_lib.c:745:Expecting: TRUSTED CERTIFICATE
> ```

This looks to me as if there is no certificate or it is somehow not trusted, and 

> For the second command I got:
> ```
> Can't open /etc/letsencrypt/live/mail.example.com/cert.pem for reading, No such file or directory
> 140581934851392:error:02001002:system library:fopen:No such file or directory:../crypto/bio/bss_file.c:69:fopen('/etc/letsencrypt/live/mail.example.com/cert.pem','r')
> 140581934851392:error:2006D080:BIO routines:BIO_new_file:no such file:../crypto/bio/bss_file.c:76:
> unable to load certificate
> ```

this looks like docker-mailserver is trying to read the certificate but can't, most likely due to a wrong mount. Have a look at #1549 too. I'd check certificates, because I think there is the error. Make sure your certificates are valid for the subdomain and mounted correctly, Just go over it again.

**Comment by aendeavor** [comment]
Please notice the **_EDIT_** in the answer above:) I misread a question.

**Comment by aendeavor** [comment]
closed due to inactivity

**Comment by xmatic** [comment]
@aendeavor : thank you very much for your suggestions and help. I took a long break from `docker-mailserver` because I was feeling burned out by it, and hence didn't respond to your suggestions earlier. Yesterday, I was able to solve the issue (thanks to your suggestions), and am posting my findings here in case it helps anyone else (or in case I need to refer to them in future).

1. As spotted by you, there was a problem with my DNS settings. I updated the DNS settings as suggested by you, and the image below contains the current, working, DNS settings. Note: my domain registrar is Hover.com.
![github mailserver 1 hover dns](https://user-images.githubusercontent.com/5103936/107136112-a453e000-68bd-11eb-9531-0fe25ea83340.jpg)
2. The next thing that I did is I added the following under **Volumes** (in docker-compose.yml): 
```      
    - /etc/letsencrypt:/etc/letsencrypt
```

This is documented in the Wiki as "mount your whole `letsencrypt` folder to `/etc/letsencrypt`". However, this could be clarified further as I've done above.

3. The next thing that I did is I added environmental variable `SSL_TYPE` in the `mailserver` service  (in docker-compose.yml) as follows:
```
    environment:
            SSL_TYPE: letsencrypt
```

4. Lastly, I installed `certbot` and installed the certificate via `$ certbot certonly --standalone -d mail.domain.tld`, then re-upped the container via `sudo docker-compose up -d`, and voila, Thunderbird didn't give me any warning identified the correct ports for imap and smtp (`mail.domain.com` for both), and was able to use SSL to login without a problem

5. My thanks to you, of course. And also to the following 2 pages that helped a lot:
https://icyphox.sh/blog/mailserver/
https://github.com/docker-mailserver/docker-mailserver/wiki/Configure-SSL
